src/templates/typescript/classes/classTemplate.ts

```diff
@@ -4,24 +4,139 @@ import DataType from "../../../model/dataTypes/DataType";
   static readonly baseType: string = "{{ metadata.parentTypeName }}";
+
   static readonly namespace: string = "{{ metadata.namespace }}";
-  static readonly typeName: string = "{{ metadata.originalTypeName }}";{{# if memberVariables }}
+
+  static readonly typeName: string = "{{ metadata.originalTypeName }}";
+
   {{# each memberVariables }}
   {{> complexMember member=this }}
 
+
+  {{ else }}
   {{/ each }}
-{{/ if }}
 {{# if dataType.primitive }}
-
   {{> primitiveParse }}
+{{ else }}
+  {{> complexParse }}
+{{/ if }}
+{{# if (eq dataType.normalizedName "Extension") }}
+
+  public static serializePrimitiveExtension(
+    primitive: Element
+  ): IElement | undefined {
+    let result: IElement | undefined;
+    if (primitive.id || primitive.extension) {
+      result = {};
+      result.id = primitive.id;
+      result.extension = primitive?.extension?.map((x) => x.toJSON());
+    }
+    return result;
+  }
+
+  public static serializePrimitiveExtensionArray(
+    primitives: Array<Element>
+  ): Array<IElement | null> | undefined {
+    const initial: Array<IElement | null> = [];
+    const result = primitives.reduce((accumulator, currentPrim) => {
+      const serializedExtension = Extension.serializePrimitiveExtension(
+        currentPrim
+      );
+      if (serializedExtension) {
+        accumulator.push(serializedExtension);
+      } else {
+        accumulator.push(null);
+      }
+      return accumulator;
+    }, initial);
+
+    if (result.find((x) => !!x)) {
+      return result;
+    }
+    return undefined;
+  }
+{{/ if }}
+{{# unless dataType.primitive }}
+
+  public toJSON(): I{{ dataType.normalizedName }} {
+    const result: I{{ dataType.normalizedName }} = {{# if parentDataType ~}}
+      super.toJSON()
+    {{~ else ~}}
+      {}
+    {{~/ if ~}};
+
+  {{# each memberVariables }}
+  {{# each choiceTypes }}
+    if (this.{{ ../variableName }} && {{ normalizedName }}.is{{ normalizedName }}(this.{{ ../variableName }})) {
+      result.{{ ../variableName }}{{ trimInterfaceName normalizedName}} = this.{{ ../variableName }}
+      {{~# unless systemType ~}}
+        {{~# if primitive ~}}
+          .value
+        {{~ else ~}}
+          .toJSON()
+        {{~/ if ~}}
+      {{~/ unless ~}};
+    {{# if primitive }}
+      result._{{ ../variableName }}{{ trimInterfaceName normalizedName}} = Extension.serializePrimitiveExtension(this.{{ ../variableName }});
+    {{/ if }}
+    }
+
   {{ else }}
+    if (this.{{ variableName }}) {
+      result.{{ variableName }} = this.{{ variableName }}
+      {{~# unless dataType.systemType ~}}
+        {{~# if isArray ~}}
+          {{~# if dataType.primitive ~}}
+            .filter(x => !!x).map(x => x.value) as typeof result.{{ variableName }}
+          {{~ else ~}}
+            .map((x) => x.toJSON())
+          {{~/ if ~}}
+        {{~ else ~}}
+          {{~# if dataType.primitive ~}}
```

Comment:
How does it work with Elements.value? I don't see this logic...
All Elements, not just Primitive types, have 'value' field which can be inlined.
Example Patient.gender is of AdministrativeGender type, but in FHIR JSON it's reflected as 
{
...
  "gender": "female", 
...
}